tests/0143-exponential_backoff.c

```diff
@@ -0,0 +1,571 @@
+                                if ((time_difference < high) &&
+                                    (time_difference > low)) {
+                                        error_code = -3;
+                                        break;
+                                }
+                                retry_count++;
+                        }
+                        previous_request_ts =
+                            rd_kafka_mock_request_timestamp(requests[i]);
+                }
+        }
+        for (i = 0; i < request_cnt; i++)
+                rd_kafka_mock_request_destroy(requests[i]);
+        rd_free(requests);
+        return error_code;
+}
+/**
+ * @brief offset_commit test
+ * We fail the request with RD_KAFKA_RESP_ERR_COORDINATOR_LOAD_IN_PROGRESS so
+ * that the request is retried with the exponential backoff. The max retries
+ * allowed is 2 for offset_commit. The RPC calls rd_kafka_buf_retry for its
+ * retry attempt so this tests all such RPCs which depend on it for retrying.
+ * The retry number of request is deterministic i.e no fresh requests are
+ * spawned on its own. Also the max retries is 2 for Offset Commit.
+ */
+void test_offset_commit(rd_kafka_mock_cluster_t *mcluster,
+                        const char *topic,
+                        rd_kafka_conf_t *conf) {
+        int16_t err = 0;
+        rd_kafka_t *consumer;
+        rd_kafka_message_t *rkm;
+        rd_kafka_topic_partition_list_t *offsets;
+        rd_kafka_topic_partition_t *rktpar;
+        SUB_TEST();
+        test_conf_set(conf, "auto.offset.reset", "earliest");
+        test_conf_set(conf, "enable.auto.commit", "false");
+
+        consumer = test_create_consumer(topic, NULL, conf, NULL);
+        test_consumer_subscribe(consumer, topic);
+        rkm = rd_kafka_consumer_poll(consumer, 10 * 1000);
+        if (rkm)
+                rd_kafka_message_destroy(rkm);
+        rd_sleep(4);
+        rd_kafka_mock_push_request_errors(
+            mcluster, RD_KAFKAP_OffsetCommit, 2,
+            RD_KAFKA_RESP_ERR_COORDINATOR_LOAD_IN_PROGRESS,
+            RD_KAFKA_RESP_ERR_COORDINATOR_LOAD_IN_PROGRESS);
+
+        offsets = rd_kafka_topic_partition_list_new(1);
+        rktpar  = rd_kafka_topic_partition_list_add(offsets, topic, 0);
+        /* Setting Offset to an arbitrary number */
+        rktpar->offset = 4;
+        /* rd_kafka_commit will trigger OffsetCommit RPC call */
+        rd_kafka_commit(consumer, offsets, 0);
+        rd_kafka_topic_partition_list_destroy(offsets);
+        rd_sleep(3);
+
+        err = helper_exponential_backoff(mcluster, RD_KAFKAP_OffsetCommit);
+
+        if (err == -3)
+                TEST_FAIL("Time difference is not respected!");
+
+        rd_kafka_destroy(consumer);
+        rd_kafka_mock_clear_requests(mcluster);
+        SUB_TEST_PASS();
+}
+
+/**
+ * @brief produce test
+ * We fail the request with RD_KAFKA_RESP_ERR_COORDINATOR_LOAD_IN_PROGRESS so
```

Comment:
You were earlier saying that no error except COORDINATOR_LOAD_IN_PROGRESS triggers just a retry for produce (though it's not a real error for Produce), and hence we should use that.

But what's the problem with RD_KAFKA_RESP_ERR_NOT_ENOUGH_REPLICAS? It triggers retry without refresh, too?